src/coins.cpp

```diff
@@ -49,10 +49,8 @@ CCoinsMap::iterator CCoinsViewCache::FetchCoin(const COutPoint &outpoint) const
             cacheCoins.erase(ret);
             return cacheCoins.end();
         }
-        if (ret->second.coin.IsSpent()) {
-            // The parent only has an empty entry for this outpoint; we can consider our version as fresh.
-            ret->second.AddFlags(CCoinsCacheEntry::FRESH, *ret, m_sentinel);
-        }
+        // GetCoin will never return true if the coin is spent
```

Comment:
This doesn't sound like it's `GetCoin`'s responsibility, I think the call-sites should be checking for whether the coins are spent or not, not the cache accessor method

Reply:
It's specifically documented that this is the behavior. Why shouldn't we be able to rely on it?

Reply:
If we need to document the behavior of a method, it suggests that we’ve acknowledged it may not be intuitive for everyone, indicating that we’ve conceded on finding a better design.
I think there's a way we can avoid the comment: by separating the two concerns of getting a value from cache and checking if it's spent or not.
Or if you insist, we can call it "GetUnspent" in which case we can also get rid of the comment and call-site asserts and keep the current behavior.
What do you think?

Reply:
What you are suggesting is basically a concept NACK on this PR. If we aren't confident that we will never get a spent coin from GetCoin, then we need to keep the logic to handle spent and FRESH coins.

Reply:
I won't block this since I'm fine with discussing/addressing any issues in later commits as well. 
I also don't yet understand the second order effects of these changes, so treat my comments as curious observations.

So my concern is that in this PR was that now we're tying `coin.IsSpent()` with retrieving values from the cache.
So if my understanding is correct we should either:
* be able to get a coin from the cache and ignore/assert whether it's spent or not - e.g. [`HaveCoin`](https://github.com/bitcoin/bitcoin/blob/master/src/coins.cpp#L21) will now return false when the cache contains it but it's spent... and if that's not possible in certain situations I would assert instead of returning the boolean.
So in that case `GetCoin` would just get it from the cache and the call-site will decide whether it should assert for `IsSpent` or branch based on it, something like this: https://github.com/bitcoin/bitcoin/blob/9db8b30788eff2b8e9c9f629a6a03886a28a3281/src/test/fuzz/coins_view.cpp#L167
* rename the method to reflect the new responsibility that it has

There's also a third possibility that I'm completely off on what this PR does - your patience is appreciated :)

Reply:
@l0rinc It seems to me like you are thinking this PR has any behavior change. Any coin retrieved from `GetCoin` will be unspent if `GetCoin` returns true. You can follow this logically from seeing the return value [here](https://github.com/bitcoin/bitcoin/blob/master/src/coins.cpp#L65) and then the only base class is `CCoinsViewDB` which does not save spent coins so cannot retrieve them in `GetCoin`.

This PR is recognizing that `FetchCoin` logically cannot get to [here](https://github.com/bitcoin/bitcoin/blob/master/src/coins.cpp#L52) with a spent coin. [This](https://github.com/bitcoin/bitcoin/blob/master/src/coins.cpp#L54) is a dead code path.

Reply:
Yeah, I understand that.

What I'm struggling with is the cache retrieval, and especially the invalidation logic, which seems to be scattered throughout the code and tightly coupled with the business logic.

Would you be open to a screen-sharing session to discuss this? It would help me better understand the constraints and likely speed up my review process as well.

Reply:
So my understanding that if a coin is already spent it's basically 
"less likely" to be in the cache (0% without reorgs, right?), so we want to get rid of it ASAP.
Given that the complete UTXO set doesn't fit into memory anymore, would it maybe make sense to have a more complex retainment heuristic for each cache layer, e.g. recency, dust amount, unspendables, time-locked, etc?

Reply:
> Given that the complete UTXO set doesn't fit into memory anymore, would it maybe make sense to have a more complex retainment heuristic for each cache layer,

Since we now have a linked list of entries, we could have another sentinel that tracks coins that have their flags cleared. Then, when doing a flush we could just remove a chunk from the bottom of the linked list (entries added least recently).

Reply:
Added https://github.com/bitcoin/bitcoin/pull/30849 to somewhat mitigate the `GetCoin` mutations in tests - hoping this will get us closer to excluding invalid tests states.